src/charm.py

```diff
@@ -6,20 +6,314 @@
+                    "http": {"url": "http://localhost:4445/admin/health/ready"},
+                },
+            },
+        }
+        return Layer(layer_config)
+
+    def _check_leader(self) -> None:
+        if not self.unit.is_leader():
+            # We can't do anything useful when not the leader, so do nothing.
+            raise CheckFailedError("Waiting for leadership", WaitingStatus)
+
+    def _apply_and_patch_resources(self) -> None:
+        try:
+            self.resource_handler.apply()
+            self.lightkube_client.patch(
+                StatefulSet,
+                self._name,
+                {
+                    "spec": {
+                        "template": {
+                            "spec": {
+                                "containers": [
+                                    {
+                                        "name": "hydra",
+                                        "env": [
+                                            {
+                                                "name": "DB_USER",
+                                                "valueFrom": {
+                                                    "secretKeyRef": {
+                                                        "name": f"{self._name}-database-secret",
+                                                        "key": "username",
+                                                    }
+                                                },
+                                            },
+                                            {
+                                                "name": "DB_PASSWORD",
+                                                "valueFrom": {
+                                                    "secretKeyRef": {
+                                                        "name": f"{self._name}-database-secret",
+                                                        "key": "password",
+                                                    }
+                                                },
+                                            },
+                                            {
+                                                "name": "DB_ENDPOINT",
+                                                "valueFrom": {
+                                                    "secretKeyRef": {
+                                                        "name": f"{self._name}-database-secret",
+                                                        "key": "endpoint",
+                                                    }
+                                                },
+                                            },
+                                            {
+                                                "name": "DSN",
+                                                "value": "postgres://$(DB_USER):$(DB_PASSWORD)@$(DB_ENDPOINT)/postgres",
+                                            },
```

Comment:
I'm not sure I understand the purpose of doing this patch when you could just set these values in the Pebble layer?

Support for Juju secrets has landed in upstream Juju, with support in `ops` landing soon too -- I'd just store these secrets as Juju secrets and inject them as env vars into the pebble layer.

Until the ops support lands for secrets (and in this early phase of dev) you could drop the secret values into a peer relation.

Patching the StatefulSet should really only be done as a last resort - it's pretty disruptive.